services/port_manager/src/main/java/com/futurewei/alcor/portmanager/processor/DataPlaneProcessor.java

```diff
@@ -44,6 +43,104 @@ private PortEntity getPortEntity(List<PortEntity> portEntities, String portId) {
+    }
+
+    private void buildL3Neighbors(PortContext context, InternalPortEntity internalPortEntity, PortEntity.FixedIp fixedIp, List<String> routerSubnetIds) throws Exception {
+        List<NeighborInfo> neighborInfos = context.getNetworkConfig().getNeighborInfos();
+
+        List<NeighborInfo> l3Neighbors = new ArrayList<>();
+        NeighborInfo localNeighborInfo = null;
+        for (NeighborInfo neighborInfo: neighborInfos) {
+            if (neighborInfo.getPortId().equals(fixedIp.getIpAddress())) {
+                localNeighborInfo = neighborInfo;
+            }else if (routerSubnetIds.contains(neighborInfo.getSubnetId()) &&
+                    !neighborInfo.getSubnetId().equals(fixedIp.getSubnetId())) {
+                l3Neighbors.add(neighborInfo);
+            }
+        }
+
+        if (localNeighborInfo == null) {
+            localNeighborInfo = getNeighborInfo(context, internalPortEntity, fixedIp);
+        }
+
+        List<NeighborEntry> neighborTable = buildNeighborTable(
+                localNeighborInfo, l3Neighbors, NeighborEntry.NeighborType.L3);
+        context.getNetworkConfig().addNeighborEntries(neighborTable);
+    }
+
+    private void buildL2Neighbors(PortContext context, InternalPortEntity internalPortEntity, PortEntity.FixedIp fixedIp) throws Exception {
+        List<NeighborInfo> neighborInfos = context.getNetworkConfig().getNeighborInfos();
+
+        List<NeighborInfo> l2Neighbors = new ArrayList<>();
+        NeighborInfo localNeighborInfo = null;
+        for (NeighborInfo neighborInfo: neighborInfos) {
+            if (neighborInfo.getPortId().equals(fixedIp.getIpAddress())) {
+                localNeighborInfo = neighborInfo;
+            }else if (neighborInfo.getSubnetId().equals(fixedIp.getSubnetId())) {
+                l2Neighbors.add(neighborInfo);
+            }
+        }
+
+        if (localNeighborInfo == null) {
+            localNeighborInfo = getNeighborInfo(context, internalPortEntity, fixedIp);
+        }
+
+        List<NeighborEntry> neighborTable = buildNeighborTable(
+                localNeighborInfo, l2Neighbors, NeighborEntry.NeighborType.L2);
+        context.getNetworkConfig().addNeighborEntries(neighborTable);
+    }
+
+    private void setNeighborInfos(PortContext context, InternalPortEntity internalPortEntity) throws Exception {
+        List<NeighborInfo> neighborInfos = context.getNetworkConfig().getNeighborInfos();
+        if (internalPortEntity.getFixedIps() == null ||
+                neighborInfos == null ||
+                internalPortEntity.getBindingHostId() == null) {
+            return;
+        }
+
+        for (PortEntity.FixedIp fixedIp: internalPortEntity.getFixedIps()) {
+            List<String> routerSubnetIds = context.getRouterSubnetIds(internalPortEntity.getVpcId());
+            if (routerSubnetIds != null && routerSubnetIds.contains(fixedIp.getSubnetId())) {
+                buildL3Neighbors(context, internalPortEntity, fixedIp, routerSubnetIds);
```

Comment:
how frequently do we go into this block to trigger buildL3Neighbors? This could be a N^2 operation.